AX: Increment/Decrement actions on spinbuttons should default to arrow key behavior if implemented

hoffmanjoshua opened the pull request:
#### 69c08ffd8fb483090d7e86b8d778d00733fdce9e
<pre>
AX: Increment/Decrement actions on spinbuttons should default to arrow key behavior if implemented
<a href="https://bugs.webkit.org/show_bug.cgi?id=300727">https://bugs.webkit.org/show_bug.cgi?id=300727</a>
<a href="https://rdar.apple.com/161667991">rdar://161667991</a>

Reviewed by Tyler Wilcock.

Spinbuttons (per the spec <a href="https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/)">https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/)</a>, are meant
to support up/down arrow key events. These event handlers are often responsible for making
sure the value of these inputs move up by a particular step amount, and other such state.

In many cases (e.g., when spinbuttons are not text inputs), we fallback to sending these key
events when incrementing/decrementing via ATs. But, when a spinbutton is a text input (in
the case of contenteditable, for example), we don&apos;t send simulated up/down arrow key events
and just insert an updated value through simulated typing.

This PR changes our behavior for spinbuttons so that we always try an arrow key event first,
and if it actually changes the value of the input, don&apos;t do anything else. This will help
make the experience more uniform between AT and keyboard events.

Test: accessibility/contenteditable-spinbutton-uses-arrow-keys.html

* LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys-expected.txt: Added.
* LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alterRangeValue):

Canonical link: <a href="https://commits.webkit.org/301689@main">https://commits.webkit.org/301689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2237dc8c148594acebaaa83434e43efa2ea6c77

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126696 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133656 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78341 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a6cbe2a8-e733-4987-bd70-c0f8f49fae7e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96410 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a9f65894-49ad-49a2-8520-920fb495b7b8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129644 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37586 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113312 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76935 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/070a6013-6968-4325-bff6-5e165aabb3d4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36472 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31496 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77053 "Built successfully") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107402 "Found 1 new API test failure: TestWebKitAPI.SiteIsolation.PostMessageWithMessagePorts (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31796 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136231 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53378 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41069 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104925 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104626 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26689 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50118 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28453 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50843 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59106 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52569 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55905 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54311 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys.html:
```diff
@@ -0,0 +1,65 @@
+</head>
+<body>
+
+<span role="spinbutton"
+     aria-valuenow="2022"
+     aria-valuemin="2000"
+     aria-valuemax="2040"
+     aria-label="Year"
+     contenteditable="true"
+     id="spinbutton">
+    2022
+</span>
```
I feel like it could be nice to specifically call out that material UI follows / followed this pattern (content editable spin button), so in 10 years when someone wonders if this test is necessary, they'll know why we added it 🙂

hoffmanjoshua replied:
Good callout, added a comment above this span!

twilco commented on LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys.html:
```diff
@@ -0,0 +1,65 @@
+</script>
+</body>
+</html>
+</html>
```
extra html tag

twilco commented on LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys.html:
```diff
@@ -0,0 +1,65 @@
+	document.getElementById("spinbutton").focus();
+	
+	let lastKnownValue = axSpinbutton.stringValue;
+	testOutput += `#spinbutton initial value: ${lastKnownValue}\n`;
```
output instead of testOutput

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1693,6 +1693,15 @@ void AccessibilityNodeObject::alterRangeValue(StepAction stepAction)
         return;
 
+    // First try a keyboard event to see if that affects the value of the spinbutton, since authors are supposed to handle up/down keys.
+    // We can't check the event listeners directly here since those may be on an ancestor or the document.
```
Since this comment specifically refers to spin buttons, it might be nice to put it inside the `if (spinbutton) { }` block.

hoffmanjoshua replied:
Moved!

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1693,6 +1693,15 @@ void AccessibilityNodeObject::alterRangeValue(StepAction stepAction)
+        float beforeValue = valueForRange();
+        postKeyboardKeysForValueChange(stepAction);
+        if (beforeValue != valueForRange())
+            return;
```
Could be nice to have a comment explaining why we return if the value changed? I feel like it's obvious considering our recent discussions, but may not be obvious 5 years from now

hoffmanjoshua committed:
AX: Increment/Decrement actions on spinbuttons should default to arrow key behavior if implemented
https://bugs.webkit.org/show_bug.cgi?id=300727
rdar://161667991

Reviewed by Tyler Wilcock.

Spinbuttons (per the spec https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/), are meant
to support up/down arrow key events. These event handlers are often responsible for making
sure the value of these inputs move up by a particular step amount, and other such state.

In many cases (e.g., when spinbuttons are not text inputs), we fallback to sending these key
events when incrementing/decrementing via ATs. But, when a spinbutton is a text input (in
the case of contenteditable, for example), we don't send simulated up/down arrow key events
and just insert an updated value through simulated typing.

This PR changes our behavior for spinbuttons so that we always try an arrow key event first,
and if it actually changes the value of the input, don't do anything else. This will help
make the experience more uniform between AT and keyboard events.

Test: accessibility/contenteditable-spinbutton-uses-arrow-keys.html

* LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys-expected.txt: Added.
* LayoutTests/accessibility/contenteditable-spinbutton-uses-arrow-keys.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alterRangeValue):

Canonical link: https://commits.webkit.org/301689@main
